utils/swap_math.go

```diff
@@ -142,68 +185,82 @@ func calcReachAmount(
+			deltaL = MulDivRoundingDown(
+				constants.Q96, new(big.Int).Mul(absDelta, fee), new(big.Int).Mul(TwoFeeUnits, currentSqrtP),
 			)
 		}
 	} else {
-		// returnedAmount = (liquidity + deltaL)/nextSqrtP - (liquidity)/currentSqrtP
-		// if exactInput, minimise actual output (<0, make less negative) so we avoid sending too much
-		// if exactOutput, maximise actual input (>0) so we get desired output amount
-		returnedAmount = new(big.Int).Add(
-			MulDivRoundingUp(new(big.Int).Add(liquidity, deltaL), constants.Q96, sqrtRatioTargetX96),
-			new(big.Int).Mul(MulDivRoundingUp(liquidity, constants.Q96, sqrtRatioCurrentX96), constants.NegativeOne),
-		)
-	}
+		// obtain the smaller root of the quadratic equation
+		// ax^2 - 2bx + c = 0 such that b > 0, and x denotes deltaL
+		a := fee
+		b := new(big.Int).Mul(new(big.Int).Sub(FeeUnits, fee), liquidity)
+		c := new(big.Int).Mul(new(big.Int).Mul(fee, liquidity), absDelta)
 
-	if exactIn && returnedAmount.Cmp(constants.One) == 0 {
-		// rounding make returnedAmount == 1
-		returnedAmount = constants.Zero
+		if isToken0 {
+			// a = feeInFeeUnits
+			// b = (FEE_UNITS - feeInFeeUnits) * liquidity - FEE_UNITS * absDelta * currentSqrtP
+			// c = feeInFeeUnits * liquidity * absDelta * currentSqrtP
+			b = new(big.Int).Sub(b, MulDiv(new(big.Int).Mul(FeeUnits, absDelta), currentSqrtP, constants.Q96))
+			c = MulDiv(c, currentSqrtP, constants.Q96)
+		} else {
+			// a = feeInFeeUnits
+			// b = (FEE_UNITS - feeInFeeUnits) * liquidity - FEE_UNITS * absDelta / currentSqrtP
+			// c = liquidity * feeInFeeUnits * absDelta / currentSqrtP
+			b = new(big.Int).Sub(b, MulDiv(new(big.Int).Mul(FeeUnits, absDelta), constants.Q96, currentSqrtP))
+			c = MulDiv(c, constants.Q96, currentSqrtP)
+		}
+
+		deltaL = GetSmallerRootOfQuadEqn(a, b, c)
 	}
 
-	return returnedAmount
+	return deltaL
 }
 
 // calcIncrementalLiquidity calculates deltaL, the swap fee to be collected for an intermediate swap step,
 // where the next (temporary) tick will be crossed
 func calcIncrementalLiquidity(
-	sqrtRatioCurrentX96, sqrtRatioTargetX96, liquidity, absAmount *big.Int, exactIn, isToken0 bool,
+	absDelta *big.Int,
+	liquidity *big.Int,
+	currentSqrtP *big.Int,
+	nextSqrtP *big.Int,
+	isExactInput bool,
+	isToken0 bool,
 ) (deltaL *big.Int) {
 	var tmp1, tmp2, tmp3 *big.Int
 
 	// this is when we reach the target, then we have target_X96
 	if isToken0 {
 		// deltaL = nextSqrtP * (liquidity / currentSqrtP +/- absDelta)) - liquidity
 		// needs to be minimum
-		tmp1 = MulDiv(liquidity, constants.Q96, sqrtRatioCurrentX96)
-		if exactIn {
-			tmp2 = new(big.Int).Add(tmp1, absAmount)
+		tmp1 = MulDiv(liquidity, constants.Q96, currentSqrtP)
+		if isExactInput {
+			tmp2 = new(big.Int).Add(tmp1, absDelta)
 		} else {
-			tmp2 = new(big.Int).Sub(tmp1, absAmount)
+			tmp2 = new(big.Int).Sub(tmp1, absDelta)
 		}
-		tmp3 = MulDiv(sqrtRatioTargetX96, tmp2, constants.Q96)
+		tmp3 = MulDiv(nextSqrtP, tmp2, constants.Q96)
```

Comment:
fix a bug here, correct code in contract is `nextSqrtP`, not `targetSqrtP`
https://github.com/KyberNetwork/ks-elastic-sc-v2/blob/1902450fd9bcbc39c8cf53b0570837513d32cdfb/contracts/libraries/SwapMath.sol#LL233C45-L233C45